{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# COMP90014 Week 4 lab "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Programming exercises: de novo assembly with De Bruijn graphs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For this week's lab you'll be writing programs to count k-mers and to find the edges of a de Bruijn graph.\n",
    "\n",
    "**If you are new to programming**, or want to learn Python, you might want to work through the more gradual tutorial exercises <a href=\"kmer_workshop.html\">here</a> before coming back to this document. These will introduce you to concepts you need to solve the problems here.\n",
    "If you're new to programming, you might find that you spend the whole lab session on the easier tutorial document. That is fine! The exercises here are incremental, so the hardest of them are really intended for experienced programmers.\n",
    "\n",
    "**If you are experienced with programming**, you can just skip straight to the exercises below, though you still might find it worth skimming through the tutorial document above. In particular it's worth checking the [Dictionaries: algorithmic considerations](kmer_workshop.html#dictionaries) section."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Setup"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You will need to be logged into a system that has Python installed, and a text editor so that you can edit your programs. One option is the linux environment on the lab servers, which you can log into as per the ssh instructions in the [Week 1 lab](https://docs.google.com/document/d/1AkIa23Gz4XVmLBLoBD6oZPKdWJdI6Q2hj95JZ1FueIk/preview).\n",
    "\n",
    "Download the starter code and the example input data from\n",
    "\n",
    "http://claresloggett.github.io/comp90014_workshops/assembly_workshop_data.zip\n",
    "\n",
    "or\n",
    "\n",
    "http://claresloggett.github.io/comp90014_workshops/assembly_workshop_data.tar.gz \n",
    "\n",
    "and unzip it.\n",
    "\n",
    "This should get you a directory called `assembly_workshop_data`, which contains some starter code that you can edit, and a set of example input data files to check your programs."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Programming challenges"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Remember that if you find these difficult or aren't sure where to start, you can try <a href=\"kmer_workshop.html\">this document</a> first (or instead).\n",
    "\n",
    "###Challenge 1\n",
    "\n",
    "Given a set of strings in a file, return all possible kmers. A starting program has been provided for you which just prints out the strings themselves, without finding the kmers. You should modify this code to print out the kmers instead. Each unique kmer should only be printed once.\n",
    "To run the starting code using the first example input, at the command line type:\n",
    "\n",
    " \tpython find_kmers.py input1a.fa 3\n",
    "\n",
    "You can try it for the various provided example input files (use `ls` to see them).\n",
    "To edit the file, you need a text editor - if you're logged in to the lab servers and not sure what editor to use, you can use nano like this:\n",
    "\n",
    "    nano find_kmers.py\n",
    "\n",
    "####Example output\n",
    "\n",
    "~~~\n",
    "$ python find_kmers.py input1b.fa 4\n",
    "HAPP\n",
    "NESS\n",
    "PINE\n",
    "APLI\n",
    "APPI\n",
    "INES\n",
    "PLIN\n",
    "INET\n",
    "\n",
    "$ python find_kmers.py input2a.fa 6\n",
    "SSIPPI\n",
    "SSISSI\n",
    "MISSIS\n",
    "~~~\n",
    "\n",
    "\n",
    "###Challenge 2\n",
    "\n",
    "Modify your program to also count how many times each kmer is seen, and print out this count together with the kmers. \n",
    "\n",
    "*NB: You might want to make a copy of your program, e.g.*\n",
    "\n",
    "    cp find_kmers.py count_kmers.py\n",
    "\n",
    "*and edit that instead, so that you don't overwrite your solution to Challenge 1.*\n",
    "\n",
    "Optionally: print out a kmer and its count only if the count is above some threshold. \n",
    "This is what is done by Velvet's *coverage cutoff* threshold.\n",
    "\n",
    "####Example output\n",
    "\n",
    "~~~\n",
    "$ python count_kmers.py input1a.fa 3\n",
    "PIN 2\n",
    "APP 2\n",
    "INE 2\n",
    "PPI 2\n",
    "ESS 1\n",
    "NES 1\n",
    "HAP 1\n",
    "\n",
    "$ python count_kmers.py input2a.fa 4\n",
    "ISSI 2\n",
    "SSIS 2\n",
    "SSIP 1\n",
    "SIPP 1\n",
    "IPPI 1\n",
    "MISS 1\n",
    "SISS 1\n",
    "~~~\n",
    "\n",
    "###Challenge 3\n",
    "\n",
    "Modify your program so that, after finding all unique kmers, it then finds edges between kmers. This means it should look for all (k-1) overlaps between the end of one kmer and the start of another. Print out the edges as node pairs, like so:\n",
    "\n",
    "    PIN INE\n",
    "    APP PPI\n",
    "    NES ESS\n",
    "    HAP APP\n",
    "    INE NES\n",
    "    PPI PIN\n",
    "\n",
    "####Example output\n",
    "\n",
    "~~~\n",
    "$ python find_edges.py input1b.fa 4\n",
    "HAPP APPI\n",
    "PINE INES\n",
    "PINE INET\n",
    "APLI PLIN\n",
    "INES NESS\n",
    "\n",
    "$ python find_edges.py input2a.fa 5\n",
    "SISSI ISSIS\n",
    "MISSI ISSIS\n",
    "SSIPP SIPPI\n",
    "ISSIS SSISS\n",
    "SSISS SISSI\n",
    "~~~\n",
    "\n",
    "**Think:** Is your method for finding edges efficient? What would happen if there were lots of kmers? If you're comparing every kmer to every other kmer, the algorithm will have $O(n^2)$, where n is the number of kmers. It's possible to do it in $O(n)$.\n",
    "\n",
    "**Optional:** drawing the graph\n",
    "When finding edges, you could just print the edges as you go, or store them in lists, or, optionally, you can use a Python library like networkx for this. To build a directed graph using networkx, try code like the following, where we add an example edge between the nodes \"HAPP\" and \"APPI\":\n",
    "\n",
    "~~~python\n",
    "import networkx\n",
    "graph = networkx.DiGraph()\n",
    "graph.add_edge(\"HAPP\",\"APPI\")\n",
    "print graph.edges()\n",
    "~~~\n",
    "\n",
    "If you build a graph usng networkx, it's also fairly easy to display it if you'd like, using networkx.draw(). Networkx will not lay out the graph very nicely unless you have Graphviz installed, but it can still be interesting to try. Try code like this:\n",
    "\n",
    "~~~python\n",
    "import networkx\n",
    "import pylab\n",
    "< ... build your de Bruijn graph ... >\n",
    "networkx.draw_spring(graph, node_size=1200, node_color=\"white\", node_shape=\"s\")\n",
    "pylab.savefig(\"graph.png\")\n",
    "~~~\n",
    "\n",
    "Building your graph using a graph structure like that provided by networkx also makes it much easier to carry out more advanced operations, such as extracting contigs.\n",
    "\n",
    "###Further work\n",
    "\n",
    "If you're interested in looking at further implementations of assembly algorithms, you might want to look at the problems in the Rosalind Project's Assembly Topic: http://rosalind.info/problems/topics/genome-assembly/."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
